source/extensions/transport_sockets/tls/context_impl.cc

```diff
@@ -979,21 +979,32 @@ OcspStapleAction ServerContextImpl::ocspStapleAction(const CertContext& cert_con
   PANIC_DUE_TO_CORRUPT_ENUM;
 }
 
+
 int ServerContextImpl::handleOcspStapling(SSL* ssl, void*) {
   const bool client_ocsp_capable = isClientOcspCapable(ssl);
-  // returns server cert selected for this connection
-  // see https://www.openssl.org/docs/man1.1.1/man3/SSL_set_tlsext_status_ocsp_resp.html
-  // for details
-  auto* selected_cert = SSL_get_certificate(ssl);
-  const auto& cert_context = certificateContext(selected_cert);
-  auto ocsp_staple_action = ocspStapleAction(cert_context, client_ocsp_capable);
+  
```

Comment:
Are you sure that iterating with `SSL_CTX_set_current_cert(ctx, SSL_CERT_SET_FIRST/NEXT)` is going to match the iteration over `tls_context_.cert_contexts_`?

Reply:
Your objection makes sense. Digging into the code, I concluded that the two collections have the same members (loaded using `SSL_CTX_use_certificate`) but they may be ordered differently. 
Now trying your suggested solution (2).

Reply:
Implementing a solution that calls: SSL_select_current_cert instead of SSL_set_current_cert